### ISSUE TITLE
2 Bugs with enhanced RST visualizer

### ISSUE THREAD

**Issue by amir-zeldes** [korpling/ANNIS#882]
Bug 1: in ANNIS4, eRST graphs are laid out incorrectly if secondary edges are present. Otherwise the graph structure seems OK, compare GUM_academic_art (no secedges, using GUM v10.2) and GUM_essay_distraction (has secedges). See the image below for an example.

Bug 2: even in RST graphs which visualize correctly, signal annotation create duplicate buttons and relation labels on the same relation (second image).

Neither of these bugs occurs in ANNIS3.

1. Import GUM 10.2
2. Search for `sec:signaled_relation` to ensure you have a secedge in the document
3. Open the RST view

**Desktop (please complete the following information):**
 - Operating System: Windows 11
 - Browser: FireFox
 - ANNIS Version: annis-4.12.3-desktop
 

![image](https://github.com/user-attachments/assets/d5e3614f-be25-4731-b6af-3a90098265e8)

![image](https://github.com/user-attachments/assets/1ee4be78-3ca4-4a83-88a4-f0f3885a4b33)

**Comment by amir-zeldes** [comment]
PS - I jut noticed the relations are also followed by either ",true" or ",false"... (second pic)

**Comment by amir-zeldes** [comment]
Hi again - are there any updates on this? I'd like to upgrade to ANNIS4.X because GUM v11 is too heavy for the old postgres backend, but unfortunately this issue prevents the upgrade.

**Comment by thomaskrause** [comment]
I try to find the difference in ANNIS 4 and 3 about the returned data, which e.g. can be a different order of edges which can influence the traversal and might trigger bugs that have been there before.

I found one inconsistency which I can't explain yet the following code the `index` is -1 because the processing information with the sentence index is missing.

https://github.com/korpling/ANNIS/blob/c4a03e3fa4604601527a0a9b0713d67d8c8165ac/src/main/java/org/corpus_tools/annis/gui/visualizers/component/rst/RSTImpl.java#L373-L380

Only `isSegment(currNode)` is checked here, but in the place where the processing annotation is created, it is also checked that this is not a signaling node with `!isSignalNode(...)`

https://github.com/korpling/ANNIS/blob/c4a03e3fa4604601527a0a9b0713d67d8c8165ac/src/main/java/org/corpus_tools/annis/gui/visualizers/component/rst/RSTImpl.java#L712-L731

The left sentence index keeps to be "undefined" or -1, but the right index is updated at some point. This might be because of the traversal and I have to check this part in more detail.

**Comment by thomaskrause** [comment]
@amir-zeldes Can you confirm that the layout bug also does not happen in ANNIS 3.7.1 (with GUM 10.0) I tried this locally and there is also a "-1-73" dangling segment without any content and the "if" structure is wrongly located, too.

![Image](https://github.com/user-attachments/assets/e8acbd37-6e9e-480f-9db1-9f185893472c)

Or maybe I don't understand the layout correctly, but these was the issue I was focussing on.

**Comment by amir-zeldes** [comment]
The image you're showing is definitely wrong, so if that's from 3.7.1 then the problem predates ANNIS4, or there's a problem with that data... I have some ANNIS 3.7.0 snapshot front end running here and it seems to run GUM 10 fine:

https://gucorpling.org/annis

It's a bit faded but you should be able to see the signals and secondary edges next to the modal About window:

![Image](https://github.com/user-attachments/assets/cc3b5643-7e55-4019-9320-9bad569bf611)


I can try to install 3.7.1 somewhere if you think that matters?

**Comment by thomaskrause** [comment]
Yes, it would help to pinpoint the version the issue was introduced. The layouting of the RST trees and the traversal is quite complex and depends on several conditions. It would have to have a "diff" of the Salt XML (which is generated internally from the relANNIS data) that produces the correct visualization with the Salt that triggers the issue. Thus, it is easier to find out what to look out for in the traversal conditions.

**Comment by amir-zeldes** [comment]
OK, I can confirm that the bug happens with the official 3.7.1 release using Windows kickstarter and PG14, testing on the official release of GUM v11.0.0:

![Image](https://github.com/user-attachments/assets/116d4e6e-c0ac-4a77-97e3-74249a99f957)

And we know that it works on 3.7.0 from the server, is that enough? I can't install a clean 3.7.0 since that artifact seems to be missing on the ANNIS releases page: 

https://github.com/korpling/ANNIS/releases?page=6

I can try to test 3.6.0, though I just noticed it doesn't like my Postgres. What version of PG is ANNIS 3.6 expecting?

**Comment by thomaskrause** [comment]
The 3.7.0 release is at https://github.com/korpling/ANNIS/releases/tag/annis-3.7.0 but does not display the correct behaviour. I think the snapshot version you used is because of a bugfix we applied to the `annis3/develop` branch, but has not been released in ANNIS 3

```patch
diff --git a/annis-service/src/main/java/annis/sqlgen/SaltAnnotateExtractor.java
b/annis-service/src/main/java/annis/sqlgen/SaltAnnotateExtractor.java
index 00ba46d79..bb13a0f54 100644
--- a/annis-service/src/main/java/annis/sqlgen/SaltAnnotateExtractor.java
+++ b/annis-service/src/main/java/annis/sqlgen/SaltAnnotateExtractor.java
@@ -946,10 +946,10 @@ public class SaltAnnotateExtractor implements
AnnotateExtractor<SaltProject>
for (Relation mirror : mirrorRelations)

{
- if (mirror != rel && mirror instanceof SRelation)
+ if (mirror != rel && mirror instanceof SDominanceRelation)
{
// check layer
- SRelation mirrorRel = (SRelation) mirror;
+ SDominanceRelation mirrorRel = (SDominanceRelation) mirror;

Set<SLayer> mirrorLayers = mirrorRel.getLayers();
if (mirrorLayers != null)
```

I just tested a locally compiled ANNIS Kickstarter from the `annis3/develop` branch and it does indeed work with this version.

So I have a working version locally to compare the generated Salt to and will further investigate. The hunch based on the bugfix in ANNIS 3 is also, that this has to do with "mirror edges". The JavaDoc on the corresponding method describes the mirror edges in more detail:

>    In ANNIS there is a special combined dominance component which has an empty
>   name, but which should not directly be included in the Salt graph.
>   
>   This functions checks if a dominance relation with empty name has a
>   "mirror" relation which is inside the same layer and between the same nodes
>   but has an relation name. If yes the original dominance relation is an
>   artificial one. The function will return true in this case and update the
>    mirror relation to include information about the artificial dominance

**Comment by amir-zeldes** [comment]
>  this has to do with "mirror edges"

I'm not 100% sure but I feel like this might have to do with the "naked" dominance operator `>`, because for Tiger trees, we wanted to be able to search for just `>edge`, `>secedge` or both (`>`). So maybe mirror edges are the 'duplicate' d edges with no edge type?

> I backported the latest state of the ANNIS 3 visualizer with minimal changes to adapt to the changes made when changing to graphANNIS

That's fantastic, thank you! Any chance I could convince you to also add back an option to display document counts in the corpora list? That would make ANNIS4 feature-complete compared to ANNIS3!
